### PR TITLE
Let `create_function` reference ECR via ImageUri

### PR DESCRIPTION
**Pull request by bkeane**
# tl;dr
The core of this PR is to allow the `create_function` method of lambda to reference the mocked ECR backend to find images. Most notably it leads to "image not found" behavior when trying to deploy with an `imageUri` which does not exist within the mocked ECR backend. It also leads to the correct retrieval of attributes such as `CodeSha256` for the `list_functions` pathways.

# Why?
While using moto to test configuration flow of lambda deployments via ECR images I noticed two things:
1. Image digests are being created non-deterministically.
2. The lambda deployment is mocking out the CodeSha256 response parameter arbitrarily by hashing the imageUri.
This is all fine and good while interacting with each service independently, but I wanted to be able to populate the ECR repo with images via `put_image` and subsequently mock deploy said images.

# What?
## Image Manifests
I added a some helper functions to a `test_ecr/test_ecr_helpers.py` file. These implement a methodology that _attempts_ to better replicate the behavior of the Registry V2 in that a manifest with all the same layers as another manifest will generate the same imageDigest. This is rather approximate. But moves things in a more same-same direction with ECR (I believe). 

I went down this road due to the original reason I was making this PR: namely to wire up the image digest parameter so I could control the image digest via the boto3 request. I still may add this back in. But for the sake of what I was trying to accomplish I realized that I needed to have the `create_function` method on the lambda side of things operate more closely to AWS.

## `create_function`
I removed the arbitrary population of the codesha256 parameter and friends, and replaced it with a call to the `ecr_backend`. Now, when `create_function` is called, and the `imageUri` case is hit, the uri is used to call the ECR backend to go find the image, which ought to be mocked there before deployment of said image.

# Considerations
This may be a breaking change for some, as it completes the mock behavior between Lambda and ECR when deploying from images. I lean towards this being a change for the good, as a mocked lambda deployment with no ECR image present in the ECR mock presents a potentially caught failure.

### CODE REVIEW

**Review by bblommers — CHANGES_REQUESTED** [review]
Hi @bkeane, thanks for submitting this PR!

The code overall looks good, but the fact that it's a breaking change worries me. Can we add this check behind a feature-flag, so that by default an unknown image is still allowed?

We always use environment variables for flags like this, so for this feature it could be something like `LAMBDA_FN_ALLOW_UNKNOWN_IMAGE=TRUE`. These env vars are then exposed/read in moto.settings.py

So the end result could look something like:
```
from moto import settings

try:
    ...
    # changes in this pr
except:
    if settings.lambda_fn_allow_unknown_image:
        ...
        # existing code, with random code_sha_256
    else:
        raise ImageNotFound
```

**Comment by bkeane** [comment]
I put this all behind a feature flag. I opted for `settings.lambda_stub_ecr` naming, am completely fine with calling it something else.

My rationale: the current behavior feels like a stub of ECR vs a mock of ECR (where, say, mock_lambda would reference mock_ecr).

I also opted to make the behavior on or off and not a fail-over situation. As I believe if folks want to verify their lambdas are deploying from images that exist, they won't want that scenario to fail-over to a stubbed / randomized set of values + OK response; they would want :boom: 

As a little "might as well" I also now have "CodeSize" being populated reasonably in the mocked case.

Any tweaks or suggestions enthusiastically welcomed. Thanks!

File: tests/test_awslambda/test_lambda.py [hunk 1]
```diff
@@ -236,6 +258,69 @@ def test_create_function_from_image():
     code.should.have.key("ResolvedImageUri").equals(resolved_image_uri)
 
 
+@mock_lambda
+def test_create_function_from_mocked_ecr(
```
Comment (bblommers) [inline tests/test_awslambda/test_lambda.py:262]: Env vars cannot (easily) be set in ServerMode, so let's just skip the two new tests there:
```
if settings.TEST_SERVER_MODE:
    raise SkipTest("..")
```
Some tests in `test_lambda_invoke.py` follow the same pattern, such as `test_invoke_lambda_using_environment_port`. 
(Just note: that test is the other way around, it _only_ runs in ServerMode - we want to _not_ run in ServerMode)

**Review by bblommers — CHANGES_REQUESTED** [review]

File: tests/test_awslambda/test_lambda.py [hunk 2]
```diff
@@ -205,18 +208,37 @@ def test_create_function__with_tracingmode(tracing_mode):
...
+        os.environ["LAMBDA_STUB_ECR"] = "TRUE"
+
+
 @mock_lambda
-def test_create_function_from_image():
+def test_create_function_from_stubbed_ecr():
     lambda_client = boto3.client("lambda", "us-east-1")
     fn_name = str(uuid4())[0:6]
-    image_uri = "111122223333.dkr.ecr.us-east-1.amazonaws.com/testlambda:latest"
+    image_uri = f"{ACCOUNT_ID}.dkr.ecr.us-east-1.amazonaws.com/testlambda:latest"
```
[the remark this answers is not in the document]
Comment (bkeane) [reply tests/test_awslambda/test_lambda.py:232]: Absolutely. Good call out.

File: moto/settings.py [hunk 3]
```diff
@@ -69,6 +69,13 @@ def allow_unknown_region() -> bool:
     return os.environ.get("MOTO_ALLOW_NONEXISTENT_REGION", "false").lower() == "true"
 
 
+def lambda_stub_ecr() -> bool:
+    # Whether to stub or mock ecr backend when deploying image based lambdas.
+    # True => don't requiring image presence in moto ecr backend for `create_function`.
+    # False => require image presence in moto ecr backend for `create_function`
+    return os.environ.get("LAMBDA_STUB_ECR", "TRUE").lower() != "false"
```
[the remark this answers is not in the document]
Comment (bblommers) [reply moto/settings.py]: The two options have a different functionality, actually. The bottom variable (`SOME_CONFIG = ..`) is resolved in the beginning, when Moto is loaded/imported. That means that the env var has to be set before everything else - if a user sets/changes the env var afterwards, Moto has no way of knowing that it was updated.

Because the first option is inside a method, the value is resolved everytime the method is called - i.e., users can change the environment variable whenever they please, and Moto will check the value of the env var every time the code is executed.

Which option is better depends on two things: how often does Moto need to know the value (i.e., how often do we need to make the call to `os.environ`), and how likely is it that the user wants to change this value halfway during the test execution.

In our scenario, we can stick to using the function. Moto only needs to check the env var once, during a specific function call, so the overhead is neglible. It is also likely that the user will want to change this value for different tests, to verify different behaviours.
